### ISSUE TITLE
"Clear browsing data" user settings

### ISSUE THREAD

**Issue by Kanichiro** [brave/browser-laptop#4082]
**Did you search for similar issues before submitting this one?**
Yes
**Describe the issue you encountered:**

**Expected behavior:**
Settings to stay On
- Platform (Win7, 8, 10? macOS? Linux distro?):
  Linux Mint 18 - Cinnamon
- Brave Version:
  See screenshot
- Steps to reproduce:
  1. From the History menu, click "Clear all Cookies and Site Data."
  2. Move the selector bars for Browser History & Download History from off to on
  3. Click Clear
- Screenshot if needed:
  see screen shots
- Any related issues:
  Everytime I select "Clear browsing data," I am forced to move the selector bars for Browser History & Download History from off to on. I tend to delete these items often, even hourly at times as I can visit a lot of news sites and I prefer to delete what these sites store on my computer.

Please allow your Users to make these settings default or not. I should be able to just select ON as my prefer setting and have them stay ON!
![default settings](https://cloud.githubusercontent.com/assets/22098577/18608798/e3bcbca8-7cb8-11e6-86fd-5310173de7e3.png)
![settings i prefer](https://cloud.githubusercontent.com/assets/22098577/18608799/e866e3aa-7cb8-11e6-91f2-c8fbadd07a10.png)
![brave version](https://cloud.githubusercontent.com/assets/22098577/18608801/eb3bb240-7cb8-11e6-9def-88ceb9b7fb73.png)

**Comment by maaz93** [comment]
@bsclifton I would like to take this up. 😄  I did some initial investigation. Seems like the state of the setting is not being persisted. Could you point me to a example where the state is being persisted and also retrieved? Will persisting the settings at https://github.com/brave/browser-laptop/blob/master/js/components/clearBrowsingDataPanel.js#L32 be sufficient?

**Comment by bsclifton** [comment]
Hi @maaz93! Let's do this :smile: 

So our persistent session contains Window settings and Application settings.  This requested feature is definitely an Application setting, since it would be global.  Here's our documented storage format (you're interested in the AppStore):
https://github.com/brave/browser-laptop/blob/master/docs/state.md

So what we could do would be to add a new item there called `clearBrowsingDataDefaults`:
```
    ...
  },
  about: {
    newtab: {
      gridLayout: string // 'small', 'medium', 'large'
    }
  },
  menu: {
    template: object // used on Windows and by our tests: template object with Menubar control
  },
  defaultBrowserCheckComplete: boolean // true to indicate default browser check is complete
  clearBrowsingDataDefaults: {
    clearBrowserHistory: boolean,
    clearDownloadHistory: boolean,
    ...
```

**Comment by bsclifton** [comment]
Of course, docs/state.md is only for documentation, so we'd need to implement code that reads from / writes to this new value.  The `js/components/clearBrowsingDataPanel.js` control that you linked to in your post is included from `js/components/main.js`:
https://github.com/brave/browser-laptop/blob/master/js/components/main.js#L982

Here, you have an opportunity (since we main.js has access to the application state access here) to read from the AppState and bind the value we get back as a property for the React control.  For example:
```
<ClearBrowsingDataPanel
  clearBrowsingDataDetail={this.props.windowState.get('clearBrowsingDataDetail')}
  onHide={this.onHideClearBrowsingDataPanel}
  defaultValues={this.props.appState.get('clearBrowsingDataDefaults')} />
```

Notice that we try to retrieve the values from the AppState (the data persisted by the AppStore).  The first time you run this, it'll be undefined since the data doesn't exist.  If you wanted to provide default values, you absolutely could inside of the clearBrowsingDataPanel control itself by adding a handler for React's `componentDidMount` and checking the bound value.  Something along the lines of:

`isClearBrowserHistoryChecked = this.props.defaultValues ? this.props.defaultValues.get('clearBrowserHistory') : true`

And then (fun part)- if the value changes at all, you'd want to create a new action (`/js/actions/appAction.js`) which updates the AppStore (`js/stores/appStore.js`). This logic would do a set on the appState, something like:

```
const defaultSettings = {
  clearBrowserHistory: action.clearBrowserHistory,
  ...
}
appState = appState.set('clearBrowsingDataDefaults')
```

**Comment by bsclifton** [comment]
@maaz93 I know I captured a lot of information (and you may already know large chunks of this), but I hope this is helpful. Please ask me any questions- I'd love to help more :smile:

**Comment by maaz93** [comment]
@bsclifton That's so clear! 😄  I think this is sufficient to start off. I had one question though. There are 3 options in the History tab, that open up the same dialog with different options selected by default. If I were to implement this default state persistence, would I have to maintain a state for each of these options or I would do something on the lines of a logical OR between the defaultPersistedState and the expected dialogState?

For eg. If I choose `Autofill data` after clicking `Clear History` and then click `Clear`. When I choose the `Clear Cache` option next, do I need to have `Autofill data` and `Cached images and files` selected by default or just `Cached images and files` ?

**Comment by bsclifton** [comment]
@maaz93 good find! This does introduce an edge case

If the user picks "Clear Cache" from the menu, it disables everything except clear cache.  We must leave it like this or else folks may accidentally erase other settings.  In these cases (where a modal is initialized with everything disabled except for X), the state of each setting should NOT persist itself (when the modal is closed, etc)

This would be a great one to write a test case for :smile: I can help with that too, when you get further along (let me know!)

**Comment by maaz93** [comment]
@bsclifton I'm a little confused with this statement

> In these cases (where a modal is initialized with everything disabled except for X), the state of each setting should NOT persist itself (when the modal is closed, etc)

There are 3 options in the browser History menu, Clear History, Clear Cache and Clear All Cookies and Site Data. And clicking on these results in the following pre-populated dialogs, respectively

Clear History
![clear_history](https://cloud.githubusercontent.com/assets/6594255/19672436/1d4e844e-9a94-11e6-9136-3af5fa06a9ea.PNG)
Clear Cache
![clear_cache](https://cloud.githubusercontent.com/assets/6594255/19672442/28818fc8-9a94-11e6-94f5-019005e6d861.PNG)
Clear All Cookies and Site Data
![clear_cookies](https://cloud.githubusercontent.com/assets/6594255/19672447/2f1f1a08-9a94-11e6-9c7b-40d6c91d649a.PNG)

In all these scenarios, some options are checked by default. In what scenario would I persist the options then?

**Comment by bsclifton** [comment]
@maaz93 interesting... so we'll have to think this through

Besides picking from the menu, you can also load this screen through preferences:
![screen shot 2016-10-24 at 10 08 40 pm](https://cloud.githubusercontent.com/assets/4733304/19673623/7805cf2c-9a36-11e6-80c1-c2dc24d734da.png)

From here, it would be great to persist the settings.  The menus though are almost describing what they should do.  "Clear Cache" should never default to also having "Saved site settings + permissions" checked for example.  Perhaps the items from the menu would be better served by having a yes/no confirm box, rather than exposing the sliders?

The version from the preferences I would feel comfortable recommending the persisting of the data.  Before we do anything, I'm going to have to defer to @bradleyrichter.  Brad, what do you think about this issue?

**Comment by bradleyrichter** [comment]
Sorry there is some confusion here.

First, we need to reset the menu options as described here: #3093

Then, with that single menu - I think it should be reset after each use because we have the automatic clearing options in the prefs security panel.


Does this make sense?

**Comment by bsclifton** [comment]
@maaz93 - you can start with #3093 and then we can figure out how to handle this one

@bradleyrichter with regards to this ticket, the feature request was to persist the settings that were chosen. You're proposing that the items which are enabled/disabled in that modal get reset with each use?  If so, we can close this ticket as a won't fix

I can see value in saving the last selected value, in case the person is clearing something often

**Comment by maaz93** [comment]
@bsclifton Cool. I'll start off with #3093 

Also, just my suggestion on this thread, I'm with @bsclifton and @Kanichiro on the fact that it seems like a useful feature to save the last selected values for the user, as I feel this menu is more accessible than the security panel. 😄

**Comment by bradleyrichter** [comment]
Ok. We can save the last settings since it is non-destructive until hitting the go button.

**Comment by maaz93** [comment]
@bsclifton I'll be working on this now as #3093 is closed. Will need help writing tests, once I'm done! 😃

**Comment by bsclifton** [comment]
@maaz93 how's it going? Let me know if you need help :smile:

**Comment by maaz93** [comment]
@bsclifton I have a working piece of code right now. But, I feel it can be simplified before submitting a PR. I feel I could benefit from another pair of eyes. Could you take a look at it if I pushed a commit to my forked repo? 😄

**Comment by bsclifton** [comment]
@maaz93 sure thing :smile: let me know your branch name and I'll check it out

**Comment by maaz93** [comment]
@bsclifton [Here it is](https://github.com/maaz93/browser-laptop/commit/97537e3e16ee03f667326362bbec53e785145148). The major issue that I'm facing is having two props, namely clearBrowsingDataDetail and defaultValues, that need to be selectively bound to the SwitchControl's checkedOn prop. Let me know your thoughts! 😄

**Comment by bsclifton** [comment]
@maaz93 I left you some comments on your commit https://github.com/maaz93/browser-laptop/commit/97537e3e16ee03f667326362bbec53e785145148

Please read through them and let me know if it makes sense.  Basically, we should remove the existing code which is saving what was chosen to the windowState (we don't want this, since your changes persist the data to the APP state, which is shared between windows).  That will resolve the problem you've found :smile:

**Comment by maaz93** [comment]
@bsclifton I went through the comments. I didn't think that I could delete the existing code in favor of my change. That makes so much more sense! We delete the code to store in the windowStore and persist the settings on clicking Clear directly to appStore. Thanks!

**Comment by maaz93** [comment]
@bsclifton I've been sitting on this bug a long time now. 🕐  When I started making some changes, I realised that I didn't get the flow of code. I spent some time reading/experimenting with React. I also understood the Flux architecture that you guys are also following. (Component -> Action -> Dispatcher -> Store -> Component). 

With this understanding, I have an important question to ask. Is it absolutely necessary for every component in Brave to extend `ImmutableComponent`? I believe this is the reason why there is so much code to update the state of the toggle buttons in clearBrowsingDataPanel on every click, as we cannot use this.state, we resort to using this.props everywhere. I see there are a few components which don't extend `ImmutableComponent`. 

If no, then I think I can cook up a pretty simple solution by not extending it and using this.state, else I'll have to just build on top the existing code and not replace it 😄

**Comment by bsclifton** [comment]
@maaz93 ImmutableComponent is just an optimization we introduced for components which don't need to track state. If you need to make use of state, please just inherit from React.Conponent 😄

**Comment by maaz93** [comment]
@bsclifton I'm finally done with my change! It's at [this branch](https://github.com/maaz93/browser-laptop/commit/bb66e3dbe897454cc9317c8e95059175780be7f1) in my fork. Please have a look if you can 😄 . I need some help writing tests now, as mentioned before. 😄  I would like to know what sort of tests I've to write and also how to fix broken tests.

**Comment by bsclifton** [comment]
@maaz93 nice, great job! :smile:  To get started with tests, I'd recommend getting familiar with running our webdriver tests. These are different than the unit tests because they are automating Brave running (in a specific way).

## getting started

Just like when you run Brave regularly, you'll want two terminals open. One you can run `npm run watch-all`. In the other, you can run `npm run uitest -- --grep="Clear Browsing Panel"`

This should kick off the automated tests which are in place for the screen which you edited (and likely you will get errors).  This is a great way to get started.

Since functionality changed, you

## writing the new test

This part is not going to be easy; but [I'd recommend reading our testing docs](https://github.com/brave/browser-laptop/blob/master/docs/tests.md) (which have links to webdriver documentation and more).  It's also helpful to look at other existing tests.

(as you read below, you may search in `clearBrowsingDataPanelTest.js` for the test `clears site settings and permissions`, which is very similar)

From a high level perspective, your plan for writing the test would be like this: make each test as if you were using the UI yourself
  - test setup; add data as needed (history entries, autofill data, etc).
  - write the test so that it will:
    - go to about:preferences#security
    - click the "Clear browser data now" button (using the CSS selector for click()), modal comes up
    - click some of the toggle switches to change the setting
    - hit clear
    - waitUntil getAppState() shows the expected data is cleared

**Comment by maaz93** [comment]
@bsclifton Thanks for the info. As suggested, I'll try and fix broken tests first. Then write the new ones!

**Comment by bsclifton** [comment]
@maaz93 how's it going? Any help needed? Let me know :smile:

**Comment by maaz93** [comment]
@bsclifton I'm almost there! I believe I've fixed the tests that I broke. I'm working on writing the new tests. I'll be adding the tests to `clearBrowsingDataPanelTest.js` itself. 😄

**Comment by maaz93** [comment]
@bsclifton Is there a way I can delete the `appState` from somewhere on the file system? I want to test the scenario where the user opens the clear browsing data panel for the first time and the default values are `undefined`.

**Comment by bsclifton** [comment]
@maaz93 sorry it's taken me so long to respond; the easiest way I can recommend would be to find your session folder (or the session file) and temporarily move it out of the way.  You're on Fedora, right? I believe the session is stored here: `~/.config/brave-development/session-store-1`.

Make sure all instances of Brave are closed and then try moving that file (rename as session-store-2 for example).  Then you can launch Brave and it'll be like your appState is fresh :smile:

**Comment by maaz93** [comment]
@bsclifton No problem. I'm on a Mac. Though the folder structure is not the same, the file name session-store-1 is the same. So I was able to find it. Thanks! 😄

**Comment by Kanichiro** [comment]
I was running when I made my original post (see screenshot).

As for the Brave browser, I installed the latest version available for Linux (0.12.15). This version has an updated History, Browsing file delete dialog box.

Unfortunately, The new version takes a minor inconvenience from the version (0.12.0) I was using and turned it into a giant mess!

Originally all I had to do is turn on a few switches, click clear and I'd be done. Now I need to turn on more switches and RESTART the bloody browser in order to clear my browsing history, cookies etc. Upon restart, the switches are back in the OFF position.  

I tend to clear my browsing history a lot, often times more than once per hour, and with this NEW and IMPROVED version I have to restart the browser every time I do this as well as turn on the switches.

Is there a way for me to get a copy of _Brave 0.12.0 for Linux_ to install? If not, I am done with Brave as a browser. I find this new version unacceptable. I am amazed it passed though QC. 

Please get someone who knows what they are doing to fix this mess! 

And again, is there a way for me to get a copy of the older version of _Brave (0.12.0) for Linux_? I liked that version a lot, and recommended it to people on the Linux Mint forums.

**Linux Mint version 18 64-bit Cinnamon desktop**
![lm version installed January 2017](https://cloud.githubusercontent.com/assets/22098577/21858211/e62dc21c-d7ed-11e6-83f1-0a06a170a589.png)

**Dialog box as opened**
![clear settings dialog box as opened](https://cloud.githubusercontent.com/assets/22098577/21858222/ef3fcd82-d7ed-11e6-9ac1-8d47c6588d3f.png)

**Changes to the switches I made before clearing data**
![user changed settings](https://cloud.githubusercontent.com/assets/22098577/21858273/1b3d10ac-d7ee-11e6-8635-ebc2e5c3aded.png)

**Do you want to Restart Now?**
![do you want to restart now](https://cloud.githubusercontent.com/assets/22098577/21858237/f89f85fc-d7ed-11e6-90a8-38fd75061002.png)

**Diaglog box After Restart**
![dialog box settings after restart](https://cloud.githubusercontent.com/assets/22098577/21858245/01ca5a08-d7ee-11e6-83a4-b59ff7526eec.png)

**Comment by bsclifton** [comment]
@Kanichiro this new version hasn't been merged to master yet and definitely hasn't been released yet... are you running from source using the branch we have, `0.13.1-branch`?  If not, then you are not using the fix.  We have released updates (as you saw, 0.12.15), but the fix has not been released.

You can always download any version you'd like from the releases page:
https://github.com/brave/browser-laptop/releases

Apologies that you've been having a bad time; I promise that there is a fix coming

**Comment by bsclifton** [comment]
@Kanichiro regarding the restart, that was introduced by @diracdeltas a while back:
https://github.com/brave/browser-laptop/commit/9f7073c21162fe6ff1ffd1d0964bb8f613806638

It's done as a work-around to clear WebRTC device IDs. I'll leave a comment [in the original issue](https://github.com/brave/browser-laptop/issues/4157) sharing that you feel it's inappropriate to prompt for restart. In the meantime, just choose NO and you'll get the same behavior as before.

**Comment by Kanichiro** [comment]
@bsclifton Thanks, I've downloaded and installed my former version of Brave (0.12.0 for Linux). 

Much, much appreciated for this link to get this older version! I think I'll keep stick with this version until you give the OK. 

ps: I kept the backup copy just in case!

**Comment by bsclifton** [comment]
Fixed with https://github.com/brave/browser-laptop/pull/6026 which was merged into branch `0.13.1-branch`.
